Bugfix Bearing/Range Pose3toPose3 Jacobians

DanMcGann opened the pull request:
Previously the Jacobians for bearing between two Pose3 objects were computed incorrectly as it neglected the Jacobian of a pose's translation wrt its tangent vector. 

The Jacobian for range between two Pose3 objects was computed correctly but in a non-canonical way. (Canonical = using OptionalJacobians for all chain rule computation) 

This commit fixes the bearing Jacobian, and provides a canonical implementation for both Pose3-Pose3 bearing and range Jacobians.

This was discovered while doing testing for multi-robot cases (where robots directly observe each other via bearing-range measurements) and had a significant and detrimental affect on results even causing GaussNewton solvers to completely diverge.

Also based on git this has been hiding for 8 years? So hopefully has not been causing issues for anyone!

Finally, I will note that the deleted comment in the test is just plain wrong since jacobians are of course compute WRT the tangent elements.

DanMcGann committed:
Bugfix Bearing/Range Pose3toPose3 Jacobians

Previously the jacobians for bearing between two Pose3 objects were
computed incorrectly as it neglected the jacobian of a pose's
translation wrt its tangent vector. The jacobian for range betwen two
Pose3 objects was computed correctly but in a non-canonical way. This
commit fixes the bearing jacobian, and provides a canonical
implementation for both Pose3-Pose3 bearing and range jacobians.

varunagrawal commented:
I don't know if the comment is wrong per se, or perhaps I am missing something. The Jacobians are computed wrt the tangent elements of the second pose (hence `numericalDerivative22`), but since the second pose is a point, it is $\in \mathbb{R}^3$ (which is its own tangent elements). There is no rotation aspect, which makes the test incomplete IMO but the comment explicitly states that.

I would really like to see an additional test that does a full $SE(3)$ to $SE(3)$ jacobian computation.

DanMcGann committed:
Update test bearingPose3toPose3

Update this unit test to actually test the pose3 to pose3 function and
jacobians.

DanMcGann commented:
You are completely right! 
I actually missed the fact that the test was actually testing the `(Point3)` signature function. 
It has now been update to test the `(Pose3)` signature function and full jacobians.

Sorry for the miss, TBH I did not run the test locally since it takes a very long time to build the tests on my local machine.
BUT I did test the new version of the test locally. So hopefully everything is squared away.

varunagrawal reviewed (approved):
LGTM. Thank you @DanMcGann!

varunagrawal commented:
> You are completely right! I actually missed the fact that the test was actually testing the `(Point3)` signature function. It has now been update to test the `(Pose3)` signature function and full jacobians.
> 
> Sorry for the miss, TBH I did not run the test locally since it takes a very long time to build the tests on my local machine. BUT I did test the new version of the test locally. So hopefully everything is squared away.

No worries. Thank you for being prompt on this and the nice fix. :)

ProfFan commented:
> Also based on git this has been hiding for 8 years?

Thank you for the fix lol, I guess it's just peeps really only use range factors (<me)
